interpreter/valid/valid.ml

```diff
@@ -329,15 +404,35 @@ let check_memop (c : context) (memop : ('t, 's) memop) ty_size get_sz at =
  *)
 
 let check_resume_table (c : context) ts2 (xys : (idx * idx) list) at =
+  let is_heap_cont_type = function
+    | VarHT (StatX x) ->
+       (try ignore(cont_type c (x @@ at)); true with _ -> false)
+    | DefHT dt ->
+       (match expand_def_type dt with
+        | DefContT _ -> true
+        | _ -> false)
+    | _ -> false
+  in
+  let extract_cont_heap_type = function
+    | VarHT (StatX x) ->
+       let ContT ht = cont_type c (x @@ at) in
+       ht
+    | DefHT dt ->
+       (match expand_def_type dt with
+        | DefContT (ContT ht) -> ht
+        | _ -> assert false)
+    | _ -> assert false
+  in
   List.iter (fun (x1, x2) ->
-    let FuncT (ts3, ts4) = as_func_tag_type (tag c x1) in
+    let FuncT (ts3, ts4) = as_func_tag_type (tag_type c x1) in
     let (_, ts') = label c x2 in
     match Lib.List.last_opt ts'  with
-    | Some (RefT (nul', DefHT (DefContT (ContT ht')))) ->
+    | Some (RefT (nul', ht)) when is_heap_cont_type ht ->
```

Comment:
You shouldn't need this side condition – it has to be a cont_type. So something like:
```
    | Some (RefT (nul', ht)) ->
      let ht' = cont_type_of_heap_type c ht in
      let ft = func_type_of_cont_type c ht in
```
where the conversion functions produce the error.

(Ideally, I'd like to simplify this whole match expression into a call to `check_stack`, analogous to `BrTable`. Something like:
```
    let FuncT (ts3, ts4) = func_type_of_tag_type c (tag c x1) in
    let ht = DefHT (...) in
    check_stack c (ts3 @ [RefT (NoNull, ht)]) (label c x) x.at
```
But the "..." will get pretty gnarly, and I'm not sure it'd actually work with declared subtyping.)